File: java/src/main/java/com/authdog/resources/ImpersonationResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * Impersonation grant operations.
 */
public final class ImpersonationResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create an impersonation helper.
     * @param clientParam Authdog client
     */
    public ImpersonationResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * List impersonation grants.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return grants envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode list(final String tenantId,
                         final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/impersonation-grants",
                null, null, JsonNode.class);
    }

    /**
     * Create an impersonation grant.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return created grant envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode create(final String tenantId,
                           final String environmentId,
                           final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/impersonation-grants",
                body, null, JsonNode.class);
    }

    /**
     * Revoke an impersonation grant.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param grantId grant ID
     * @return revoke envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode revoke(final String tenantId,
                           final String environmentId,
                           final String grantId)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/impersonation-grants/" + grantId
                        + "/revoke",
                null, null, JsonNode.class);
    }
}

File: csharp/Resources/ImpersonationResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Impersonation-grant Wave 3 operations
    /// </summary>
    public class ImpersonationResource
    {
        private readonly AuthdogClient _client;

        public ImpersonationResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<JObject> ListAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/impersonation-grants");

        public JObject List(string tenantId, string environmentId) =>
            ListAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> CreateAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/impersonation-grants",
                body);

        public JObject Create(string tenantId, string environmentId, object body) =>
            CreateAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> RevokeAsync(string tenantId, string environmentId, string grantId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/impersonation-grants/{grantId}/revoke");

        public JObject Revoke(string tenantId, string environmentId, string grantId) =>
            RevokeAsync(tenantId, environmentId, grantId).GetAwaiter().GetResult();
    }
}
